### PR TITLE
[MM-47966] Improve accessibility for profile popovers

### PR DESCRIPTION
**Pull request by cyrilzhang-mm**
#### Summary

What's in this PR:
- At mentions are now buttons. They can be tabbed to via keyboard, and clicked with a keyboard to open the profile popover.
- Various accessibility improvements to the profile popover. I changed most links to buttons with a link style, and I added a focus trap. The popover is now an ARIA dialog, so screen readers are also trapped. It currently doesn't have a close button, so the escape key is needed to close the popover when using screen readers. However, in the profile popover UI update, a close button will be added, so this will not be an issue.
    + Currently, focus doesn't wrap around in the popover. It might be better to make it wrap around, but it's currently difficult to determine what is the last focusable element in the popover. I think this would be easier to implement in the UI update, so we can do it then. Alternatively I can add a generic way of doing this that uses `querySelectorAll` to locate all children which might be tabbable. That is what focus trap libraries do. 
    + When the profile popover is open, users are prevented from using the up and down keys to switch between posts.
    + However, a decision I made was to automatically close the profile popover if the user starts typing and allow the focus to go to the textbox. This is because the current behaviour is that if the user starts typing, the popover won't close but focus automatically goes to the main textbox. I thought that the ability to immediately start typing a message from anywhere in the app is an intentional feature, and if I changed the behaviour to prevent the focus from going to the textbox, this could be a breaking change for users who are used to the current behaviour. But let me know if you disagree with this approach.
- All buttons that trigger profile popovers are actually buttons and properly focus.
- A new custom focus event for the a11y-controller. This allows other components to manually dispatch a focus event that will be highlighted by the controller. This was needed to allow the profile popover to focus upon opening and return focus to the triggering element upon closing.
    + If a modal is opened from the profile popover and then closed, focus also returns to the element that triggered the popover. This is the most logical element to focus.
- I tested with VoiceOver on MacOS.

I will add further comments in this PR to explain why I made certain changes to the code.

#### Ticket Link

[Jira ticket](https://mattermost.atlassian.net/browse/MM-47966)

#### Screenshots

https://user-images.githubusercontent.com/112951043/202022367-8baed026-5a66-4fbf-b534-9171defca1ea.mov

#### Release Note

```release-note
Keyboard and focus handling was improved in profile popovers and at-mentions.
```

### CODE REVIEW

File: components/add_user_to_channel_modal/add_user_to_channel_modal.tsx
Comment (cyrilzhang-mm) [inline components/add_user_to_channel_modal/add_user_to_channel_modal.tsx:29]: This change needed to be made due how the `ModalController` works. The controller passes an `onHide` function that doesn't call the `onHide` handler passed by the parent, but it does for `onExited`. However we can't call both `onHide` and `onExited` passed by the controller as it bypasses `onExited`.

File: components/profile_popover/profile_popover.tsx
Comment (cyrilzhang-mm) [inline components/profile_popover/profile_popover.tsx:79]: I'm not using this right now, but I will need to in the user groups popover.

File: components/profile_popover/profile_popover.tsx
Comment (cyrilzhang-mm) [inline components/profile_popover/profile_popover.tsx:802]: The static header is actually the first thing that gets focused inside the profile popover, rather than the first button/link. This is perfectly fine for accessibility, but after the UI update we will want to focus the close button first.

File: components/widgets/users/avatar/avatar.tsx
Comment (cyrilzhang-mm) [inline components/widgets/users/avatar/avatar.tsx:14]: I needed this to style buttons containing avatars, so they would be the right size. Let me know if there's a preferable way to doing this, though.

File: sass/components/_links.scss
Comment (cyrilzhang-mm) [inline sass/components/_links.scss:16]: I thought it could be useful to have a global style for buttons that should look like links.

File: utils/a11y_controller.js
Comment (cyrilzhang-mm) [inline utils/a11y_controller.js:26]: The difference between these two is that the a11y-controller's default behaviour is to only show a visible focus if the user is literally in the middle of using the keyboard (i.e. the user is pressing a key on the keyboard). But we want to be able to manually focus elements even if no keys are currently pressed. So, we need a distinction between these two actions.

Comment (BenCookie95) [reply utils/a11y_controller.js:26]: Might be good to add this as a comment

File: sass/components/_mentions.scss
Comment (BenCookie95) [inline sass/components/_mentions.scss:33]: This colour doesn't seem right. See screenshot of sysadmin mention below 
Screen Shot 2022-11-21 at 10 02 42 AM

**Review by BenCookie95 — CHANGES_REQUESTED** [review]
Looks really good, some minor comments

> Currently, focus doesn't wrap around in the popover.
Is this preventing you from looping through the tabbing in the popover? When you reach the last element in the popover it stops tabbing instead of going back to the first element

**Comment by cyrilzhang-mm** [comment]
>  Is this preventing you from looping through the tabbing in the popover? When you reach the last element in the popover it stops tabbing instead of going back to the first element

Yes, that's what I meant. 

Basically in order to loop the tabbing I would have to do either of:
1. Determine which elements are the first and last tabbable elements before/while rendering then manually handle the focus when the first and last elements are reached. 
2. Write a generic `querySelectorAll` to find all the tabbable elements after rendering then manually handle the focus when the first and last elements of this array are reached.

Approach 1 is currently too hard but I was thinking it would be doable after the profile popover UI update and it would be more elegant than approach 2. But Approach 2 would work too though.

**Comment by BenCookie95** [comment]
> > Is this preventing you from looping through the tabbing in the popover? When you reach the last element in the popover it stops tabbing instead of going back to the first element
> 
> Yes, that's what I meant.
> 
> Basically in order to loop the tabbing I would have to do either of:
> 
> 1. Determine which elements are the first and last tabbable elements before/while rendering then manually handle the focus when the first and last elements are reached.
> 2. Write a generic `querySelectorAll` to find all the tabbable elements after rendering then manually handle the focus when the first and last elements of this array are reached.
> 
> Approach 1 is currently too hard but I was thinking it would be doable after the profile popover UI update and it would be more elegant than approach 2. But Approach 2 would work too though.

I'm ok to wait until the popover UI update is merged and then evaluating our options

File: sass/components/_mentions.scss
Comment (BenCookie95) [reply sass/components/_mentions.scss:33]: Ah you are correct, I got mixed up with the theming

**Review by BenCookie95 — APPROVED** [review]

File: components/at_mention/at_mention.tsx
Comment (nevyangelova) [inline components/at_mention/at_mention.tsx:94]: nit: we can add null here or pass the condition in another way, as this updates the snapshots and renders an empty string classname.

**Review by nevyangelova — APPROVED** [review]
LGTM!

**Review by furqanmlk — APPROVED** [review]
Working fine

**Pull request merged**
